nomos-libp2p/src/behaviour/nat/inner.rs

```diff
@@ -269,3 +269,207 @@ where
+
+        fn listen_on(
+            &mut self,
+            id: ListenerId,
+            addr: Multiaddr,
+        ) -> Result<(), TransportError<Self::Error>> {
+            self.inner.listen_on(id, addr)
+        }
+
+        fn remove_listener(&mut self, id: ListenerId) -> bool {
+            self.inner.remove_listener(id)
+        }
+
+        fn dial(
+            &mut self,
+            addr: Multiaddr,
+            opts: DialOpts,
+        ) -> Result<Self::Dial, TransportError<Self::Error>> {
+            if !self.has_blocked_once {
+                self.has_blocked_once = true;
+                return Err(TransportError::MultiaddrNotSupported(addr));
+            }
+
+            self.inner.dial(addr, opts)
+        }
+
+        fn poll(
+            mut self: Pin<&mut Self>,
+            cx: &mut Context<'_>,
+        ) -> Poll<libp2p::core::transport::TransportEvent<Self::ListenerUpgrade, Self::Error>>
+        {
+            Pin::new(&mut self.inner).poll(cx)
+        }
+    }
+
+    #[derive(NetworkBehaviour)]
+    struct Server {
+        autonat: server::Behaviour<OsRng>,
+        identify: identify::Behaviour,
+    }
+
+    #[derive(NetworkBehaviour)]
+    struct Client {
+        nat: InnerNatBehaviour<StdRng, TestMapper, SystemGatewayDetector>,
+        identify: identify::Behaviour,
+    }
+
+    fn create_memory_swarm<Behaviour, BehaviourFn>(
+        use_blocking: bool,
+        behaviour_fn: BehaviourFn,
+    ) -> libp2p::Swarm<Behaviour>
+    where
+        Behaviour: NetworkBehaviour,
+        BehaviourFn: FnOnce(&Keypair) -> Behaviour,
+    {
+        let keypair = Keypair::generate_ed25519();
+
+        let transport = if use_blocking {
+            FirstDialBlockingTransport::default()
+                .upgrade(Version::V1)
+                .authenticate(plaintext::Config::new(&keypair))
+                .multiplex(yamux::Config::default())
+                .boxed()
+        } else {
+            MemoryTransport::default()
+                .upgrade(Version::V1)
+                .authenticate(plaintext::Config::new(&keypair))
+                .multiplex(yamux::Config::default())
+                .boxed()
+        };
+
+        libp2p::SwarmBuilder::with_existing_identity(keypair)
+            .with_tokio()
+            .with_other_transport(|_| transport)
+            .expect("transport should be supported")
+            .with_behaviour(behaviour_fn)
+            .expect("behaviour should be supported")
+            .build()
+    }
+
+    fn create_server_with_blocking_transport() -> (libp2p::Swarm<Server>, Multiaddr) {
+        let mut swarm = create_memory_swarm(true, |keypair| Server {
+            autonat: server::Behaviour::new(OsRng),
+            identify: identify::Behaviour::new(identify::Config::new(
+                "/test/1.0.0".to_owned(),
+                keypair.public(),
+            )),
+        });
+
+        let server_addr: Multiaddr = "/memory/3000".parse().unwrap();
```

Comment:
There's a `listen` function on the `libp2p_swarm_test` crate which automatically picks a random port to start listening on, maybe we can do that here? In case we were to use the same port elsewhere.

Reply:
For example: https://github.com/logos-co/nomos/blob/580c6de658eb0c9dc5c2660fba72571019d678a1/nomos-services/blend/src/core/backends/libp2p/tests/message_handling.rs#L31

Reply:
You'd need to add a `with_memory_address_external` , e.g., https://github.com/logos-co/nomos/blob/a5b817c8c61519b5f974334138b75639a5f0fc86/nomos-blend/network/src/core/with_core/behaviour/tests/bootstrapping.rs#L27.